.github/workflows/openstack/prune_openstack

```diff
@@ -1,84 +1,93 @@
 
-remove_stale_keys () {
-    echo "HAMMERTIME: $HAMMER_TIME"
-    while IFS="," read -r ID CREATED_DATE NAME
-    do
-        if [[ $CREATED_DATE < $HAMMER_TIME ]];
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack keypair delete "$ID"
-        fi
-    done < <(cat keyid_created_name.txt)
-}
+# Static constants
+use constant {
+    OS_AUTH_TYPE            => 'v3applicationcredential',
+    OS_AUTH_URL             => 'https://keystone.hou-01.cloud.prod.cpanel.net:5000/v3',
+    OS_IDENTITY_API_VERSION => '3',
+    OS_REGION_NAME          => 'RegionOne',
+    OS_INTERFACE            => 'public',
+    KEY_NAME                => "deletethis",
+    VM_NAME                 => "elevate.github.cpanel.net",
+};
 
-get_volumes () {
-    rm volume_ids.txt > /dev/null 2>&1
-    openstack volume list | grep packer  | grep available | awk '{print $2}' >> volume_ids.txt
-}
+my $openstack_path = "/usr/bin/openstack" if -x "/usr/bin/openstack";
+die "No Openstack CLI binary installed." unless $openstack_path;
+
+# Define "two hours ago" timestamp
+my $time_offset = 7200;
+my $hammer_time = POSIX::strftime( "%Y-%m-%dT%H:%M:%SZ", gmtime( time - $time_offset ) );
+
+# Run removal functions
+remove_stale_entities( 'keys',      KEY_NAME, 'keypair' );
+remove_stale_entities( 'instances', VM_NAME,  'server' );
+
+# Fetch and format entities
+sub get_entities {
+    my ($entity_type) = @_;
+
+    my $entities = {};
 
-remove_volumes () {
-    while IFS= read -r ID; do
-        openstack volume delete "$ID" || true
-    done < <(cat volume_ids.txt)
+    my ( $list_command, $show_command );
+    if ( $entity_type eq 'keys' ) {
+        $list_command = "$openstack_path keypair list -f json";
+        $show_command = "$openstack_path keypair show -f json";
+    }
+    else {
+        $list_command = "$openstack_path server list -f json --no-name-lookup";
+        $show_command = "$openstack_path server show -f json";
+    }
+
+    my $out  = _run_cmd( $list_command, 0 );
+    my $json = JSON::PP->new->decode($out);
+
+    foreach my $entry (@$json) {
+        my $entity_name = length $entry->{ID} ? $entry->{ID} : $entry->{Name};
+        my $show_out    = _run_cmd( qq{$show_command "$entity_name"}, 0 );
+        my $entity_data = JSON::PP->new->decode($show_out);
+
+        $entities->{$entity_name} = {
+            created_on => $entity_data->{created_at},
+            name       => $entity_data->{name},
+        };
+        if ( length $entity_data->{id} ) {
+            $entities->{$entity_name}->{id} = $entity_data->{id};
+        }
+    }
+    return $entities;
 }
 
-cleanup () {
-    rm -fv id_created_name.txt
-    rm -fv keyid_created_name.txt
+# Remove stale entities
+sub remove_stale_entities {
+    my ( $entity_type, $name_pattern, $delete_command ) = @_;
+
+    my $entities = get_entities($entity_type);
+
+    foreach my $entity ( grep { $_->{name} =~ /${\$name_pattern}/ } values %{$entities} ) {
```

Comment:
Can we simplify this down further?

Maybe:
```
foreach my $entity ( keys $entities ) {
    next unless $entity->{name}  =~ /${\$name_pattern}/;
```

Reply:
What If I write a tidbit explaining how simple the logic really is?

Reply:
OK Simplified.